core/shard_thread.go

```diff
@@ -0,0 +1,129 @@
+	"sync"
+	"time"
+
+	"github.com/dicedb/dice/core/auth"
+	"github.com/dicedb/dice/core/ops"
+
+	"github.com/dicedb/dice/config"
+	"github.com/dicedb/dice/core/diceerrors"
+	"github.com/dicedb/dice/server/utils"
+)
+
+type ShardID int8
+
+type ShardError struct {
+	shardID ShardID // shardID is the ID of the shard that encountered the error
+	err     error   // err is the error that occurred
+}
+
+type ShardThread struct {
+	id               ShardID                            // id is the unique identifier for the shard.
+	store            *Store                             // store that the shard is responsible for.
+	ReqChan          chan *ops.StoreOp                  // ReqChan is this shard's channel for receiving requests.
+	workerMap        map[string]chan *ops.StoreResponse // workerMap maps workerID to its unique response channel
+	workerMutex      sync.RWMutex                       // workerMutex is the workerMap's mutex for thread safety.
+	errorChan        chan *ShardError                   // errorChan is the channel for sending system-level errors.
+	lastCronExecTime time.Time                          // lastCronExecTime is the last time the shard executed cron tasks.
+	cronFrequency    time.Duration                      // cronFrequency is the frequency at which the shard executes cron tasks.
+}
+
+// NewShardThread creates a new ShardThread instance with the given shard id and error channel.
+func NewShardThread(id ShardID, errorChan chan *ShardError) *ShardThread {
+	return &ShardThread{
+		id:               id,
+		store:            NewStore(),
+		ReqChan:          make(chan *ops.StoreOp, 1000),
+		workerMap:        make(map[string]chan *ops.StoreResponse),
+		errorChan:        errorChan,
+		lastCronExecTime: utils.GetCurrentTime(),
+		cronFrequency:    config.ShardCronFrequency,
+	}
+}
+
+// Start starts the shard thread, listening for incoming requests.
+func (shard *ShardThread) Start(ctx context.Context) {
+	ticker := time.NewTicker(shard.cronFrequency)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case op := <-shard.ReqChan:
+			shard.processRequest(op)
+		case <-ticker.C:
+			shard.runCronTasks()
+		case <-ctx.Done():
+			shard.cleanup()
+			return
+		}
+	}
+}
+
+// runCronTasks runs the cron tasks for the shard. This includes deleting expired keys.
+func (shard *ShardThread) runCronTasks() {
+	DeleteExpiredKeys(shard.store)
+	shard.lastCronExecTime = utils.GetCurrentTime()
+}
+
+func (shard *ShardThread) registerWorker(workerID string, workerChan chan *ops.StoreResponse) {
+	shard.workerMutex.Lock()
+	shard.workerMap[workerID] = workerChan
+	shard.workerMutex.Unlock()
+}
+
+func (shard *ShardThread) unregisterWorker(workerID string) {
+	shard.workerMutex.Lock()
+	delete(shard.workerMap, workerID)
+	shard.workerMutex.Unlock()
+}
+
+// processRequest processes a Store operation for the shard.
+func (shard *ShardThread) processRequest(op *ops.StoreOp) {
+	response := shard.executeCommand(op)
+
+	shard.workerMutex.RLock()
```

Comment:
What's the use of locks here?
Won't it make processing slow?

Reply:
Worker map is dynamic, it is changed every time a new client connection comes up. This registration flow needs to be synchronised with query processing.